flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/ZeroCopyNettyMessageDecoder.java

```diff
@@ -0,0 +1,280 @@
+        this.intermediateState = new MessageDecodeIntermediateState();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        frameHeaderBuffer = ctx.alloc().directBuffer(NettyMessage.FRAME_HEADER_LENGTH);
+        messageHeaderBuffer = ctx.alloc().directBuffer(INITIAL_MESSAGE_HEADER_BUFFER_LENGTH);
+
+        decodeStep = DecodeStep.DECODING_FRAME_HEADER;
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        super.channelInactive(ctx);
+
+        if (intermediateState.messageHeaderParseResult != null) {
+            Buffer buffer = intermediateState.messageHeaderParseResult.getTargetDataBuffer();
+
+            if (buffer != null) {
+                buffer.recycleBuffer();
+            }
+        }
+
+        clearState();
+
+        frameHeaderBuffer.release();
+        messageHeaderBuffer.release();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (!(msg instanceof ByteBuf)) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+
+        ByteBuf data = (ByteBuf) msg;
+
+        try {
+            while (data.readableBytes() > 0) {
+                if (decodeStep == DecodeStep.DECODING_FRAME_HEADER) {
+                    ByteBuf toDecode = cumulateBufferIfNeeded(frameHeaderBuffer, data, FRAME_HEADER_LENGTH);
+
+                    if (toDecode != null) {
+                        decodeFrameHeader(toDecode);
+
+                        decodedBytesOfCurrentStep = 0;
+                        decodeStep = DecodeStep.DECODING_MESSAGE_HEADER;
+                    }
+                }
+
+                if (decodeStep == DecodeStep.DECODING_MESSAGE_HEADER) {
+                    ByteBuf toDecoder = cumulateBufferIfNeeded(messageHeaderBuffer, data, intermediateState.messageHeaderLength);
+
+                    if (toDecoder != null) {
+                        intermediateState.messageHeaderParseResult = messageParser.parseMessageHeader(intermediateState.msgId, toDecoder);
+
+                        if (intermediateState.messageHeaderParseResult.getDataBufferAction() ==
+                                NettyMessageParser.DataBufferAction.NO_DATA_BUFFER) {
+                            ctx.fireChannelRead(intermediateState.messageHeaderParseResult.getParsedMessage());
+                            clearState();
+                        } else {
+                            decodedBytesOfCurrentStep = 0;
+                            decodeStep = DecodeStep.DECODING_BUFFER;
+                        }
+                    }
+                }
+
+                if (decodeStep == DecodeStep.DECODING_BUFFER) {
+                    readOrDiscardBufferResponse(data);
+
+                    if (decodedBytesOfCurrentStep == intermediateState.messageHeaderParseResult.getDataBufferSize()) {
```

Comment:
If we return a boolean value in above `readOrDiscardBufferResponse` to indicate whether the full `BufferResponse` is got, then we can discard the global var `decodedBytesOfCurrentStep` completely, also refer to comment https://github.com/apache/flink/pull/7368/files#r376207885

Reply:
I think the main difficulty is how to maintain the discarded bytes when we want to discard the current buffer. Currently it seems a variable is still needed for bookkeeping.

Reply:
The key issue here is how to define the role of decoder. One option is let decoder handle all the logics like this PR did. Another option is let decoder only parsing the frame header as previous netty decoder did, then make all the specific message header/body parsing delegate by individual message. If so, the decoder would be very light-weight and less states to judge different steps and message types. Let's continue thinking this issue while reviewing the codes.